vars/terraform.groovy

```diff
@@ -0,0 +1,168 @@
+
+  Map parallelStages = [failFast: false]
+
+  final String sharedToolsSubDir = '.shared-tools'
+  final String makeCliCmd = "make --directory=${sharedToolsSubDir}/terraform/"
+
+  // Defines a cron trigger only if it's requested by the user through input parameter
+  if (isBuildOnProductionBranch && !isBuildOnChangeRequest) {
+    properties([
+      pipelineTriggers([defaultConfig.cronTriggerExpression])
+    ])
+  }
+
+  withEnv([
+    // https://www.terraform.io/docs/cli/config/environment-variables.html#tf_in_automation
+    "TF_IN_AUTOMATION=1",
+    // https://www.terraform.io/docs/cli/config/environment-variables.html#tf_input
+    "TF_INPUT=0",
+  ]) {
+
+
+    if (!isBuildCauseUser) {
+      parallelStages['staging'] = {
+        stage('Staging') {
+          agentTemplate(finalConfig.agentContainerImage, {
+            withCredentials(finalConfig.stagingCredentials) {
+              stage('🔎 Validate Terraform for Staging Environment') {
+                getInfraSharedTools(sharedToolsSubDir)
+
+                sh makeCliCmd + ' validate'
+              }
+              if (finalConfig.runCommonTests) {
+                stage('✅ Commons Test Terraform Project') {
+                  sh makeCliCmd + ' common-tests'
+                }
+              }
+              if (finalConfig.runTests) {
+                stage('✅ Specific Tests Terraform Project') {
+                  sh 'make -C ./tests'
+                }
+              }
+            }
+          })
+        }
+      }
+    }
+
+    parallelStages['production'] = {
+      stage('Production') {
+        agentTemplate(finalConfig.agentContainerImage, {
+          withCredentials(defaultConfig.productionCredentials) {
+            stage('🦅 Generate Terraform Plan') {
+              // When the job is triggered by the daily cron timer, then the plan succeed only if there is no changes found (e.g. no config drift)
+              // For all other triggers, the plan succeed either there are changes or not
+              String tfCliArsPlan=''
+              final String planFileName = 'terraform-plan-for-humans.txt'
+              if (isBuildCauseTimer) {
+                tfCliArsPlan='-detailed-exitcode'
+              }
+              withEnv([
+                // https://www.terraform.io/docs/cli/config/environment-variables.html#tf_cli_args-and-tf_cli_args_name
+                "TF_CLI_ARGS_plan=${tfCliArsPlan}",
+                "PLAN_FILE_NAME=${planFileName}",
+              ]) {
+                getInfraSharedTools(sharedToolsSubDir)
+
+                try {
+                  sh makeCliCmd + ' plan'
+                }
+                finally {
+                  archiveArtifacts planFileName
+                }
+              }
+            }
+
+            if (isBuildOnChangeRequest) {
+              stage('🗣 Notify User on the PR ') {
+                final String planFileUrl = "${env.BUILD_URL}artifact/terraform-plan-for-humans.txt"
```

Comment:
Why the `final` here?

Reply:
because is not expected to change after being define. You can see this as "read only". It helps the JVM GC to keep track of the correct memory blocks.